lib/gold_miner/slack/client.rb

```diff
@@ -9,8 +9,8 @@ class Slack::Client
 
     extend Dry::Monads[:result]
 
-    def self.build(api_token:, slack_client: ::Slack::Web::Client)
-      client = new(api_token, slack_client)
+    def self.build(api_token:, slack_client: ::Slack::Web::Client, author_config: AuthorConfig.default)
```

Comment:
It feels a bit strange to do this inside the Slack::Client class.

Reply:
Does it feel like `Slack::Client` knows too much about your domain?
Should it also know about the `"rupee-gold"` emoji, your keywords `"TIL"`, `"tip"`?

Just rubber-ducking :)

Reply:
> Should it also know about the "rupee-gold" emoji, your keywords "TIL", "tip"?

Those bother me less than the `AuthorConfig`, because at least they're related to a slack message. We do have a [`MessagesQuery`](https://github.com/thoughtbot/gold_miner/blob/b3ddc3476938bf3762887c1d5e21138780b2bcb9/lib/gold_miner/messages_query.rb) class, which is a builder for a Slack message search query.

If we could extract the ["Interesting Messages" concept](https://github.com/thoughtbot/gold_miner/blob/b3ddc3476938bf3762887c1d5e21138780b2bcb9/lib/gold_miner/messages_query.rb) somehow, that could be useful.

Maybe a new class?

```rb
# I'm not a fan of this name, but this is just to spark an idea
class InterestingMessagesQueries
  def initialize(slack_channel)
    @slack_channel = slack_channel
  end

  def til_messages_query = interesting_messages.with_topic("TIL")

  def tip_messages = interesting_messages.with_topic("tip")

  def golden_messages = interesting_messages.with_reaction("rupee-gold")

  private

  def interesting_messages
    MessagesQuery
      .new
      .on_channel(@slack_channel)
      .sent_after_last_friday
  end
end
```

Reply:
Good points. I've also been thinking about the "layers" of the app. Something like this might make sense:

- Client objects: SDK or wrappers around SDK code: `Slack::Client`, `OpenAi::Client`. We could think of them as part of the "infrastructure layer"

- Query and Service objects: part of the "domain layer". They use Client objects to do their thing, and return raw data, or rich objects, that are part of our domain.

- View objects: part of the "presentation layer". `OpenAiWriter`, `SimpleWriter`. They shouldn't know about Client objects. They would receive domain data and present it.

- `GoldMiner` object, and any other object we need to coordinate execution: part of the "application layer", the way through which we interact with the application.

That's my Rails brain thinking :D

Reply:
That's useful. I'd love to take some time to investigate this together and do some refactoring.

Reply:
Sure, any time!